### DISCUSSION TITLE
add-advice on magit-checkout failure

### DISCUSSION THREAD

**Discussion by ville-h** [magit/magit#4660] [Support] [answered]
I am trying to `add-advice` on `magit-checkout` with following:
```
(defun ville-magit-update-modeline-advice(&optional args)
    (interactive)
    (message "(ville-magit-update-modeline-advice) begin")
    (dolist
        (b (buffer-list))
        (with-current-buffer b
            (when buffer-file-name (vc-refresh-state))
        )
    )
    (message "(ville-magit-update-modeline-advice) end")
)


(defun ville-magit-mode-config()
    (message "(ville-magit-mode-config) begin")
    (advice-add 'magit-checkout :after #'ville-magit-update-modeline-advice)
    (message "(ville-magit-mode-config) end")
)
(add-hook 'magit-mode-hook #'ville-magit-mode-config)
```
And afterwards I execute `magit` from source buffer under version control and then `b` `b` to execute `magit-checkout`. As far as I can tell that's the function which gets called for that shortcut by default.

### what behavior you expected
I was hoping my advice would be run after `magit-checkout`

### what behavior you observed
There is an error:
```
Debugger entered--Lisp error: (wrong-number-of-arguments #<subr magit-checkout> 0)
  #<subr magit-checkout>()
  apply(#<subr magit-checkout> nil)
  magit-checkout()
  funcall-interactively(magit-checkout)
  command-execute(magit-checkout)
```

Magit version: Magit 20220412.2029 [>= 3.3.0-git], Git 2.35.2, Emacs 28.1, gnu/linux

**Comment by tarsius** [comment, marked as answer]
1. `magit-checkout` is very short. It would be best to modify a copy.
2. I am not sure if it is even possible to advice the interactive use  of a function. I would have to investigate that too, but since I would not to do that myself and instead would define a new command, I am not interested in doing that.
3. This is not a bug in Magit, and you shouldn't report it as such. Either this is not possible at all or you are doing it wrong.

**Reply by ville-h** [reply]
Well if that's considered unfashionable, how about magit providing way to `add-hook` to run user actions after magit completes theirs?

**Reply by tarsius** [reply]
Magit has many commands and I cannot add a hook to each one of them.

**Reply by ville-h** [reply]
Just looking at this particular use-case it would be the actions that change the repo's branch state. Adding a hook for them as a group would do, rather than adding a hook for each. Whether there is need for hooks for other actions should be investigated on their own merit as the need rises.

**Reply by tarsius** [reply]
You could use `magit-post-refresh-hook` for that if you through a `(when (memq this-command ...) ...)` in there.

But I recommend you don't use `vc-refresh-state` at all, see https://github.com/magit/magit/wiki/Tips-and-Tricks#updating-vcs-mode-line-information.
